jlab-sensing/ENTS-node-firmware
# Recover buffer position after reset

**Name/Affiliation/Title**
Ella Ferraz/UCSC

**Purpose of the PR**
Refer to issue #69 

**Development Environment**
Ubuntu 20.04.6 LTS
PlatformIO Core, version 6.1.16
Hardware: 2.2.1

**Test Procedure**
To verify the functionality of data retrieval from FRAM after power outages, follow these steps:

1. Run `example_retrieve_data`

- Upload and run `example_retrieve_data` on the STM32 board. This will initialize the system and write sample data to the FRAM buffer.

2. Simulate Power Outage

- After the initial setup, manually remove power from the STM32 board to simulate an unexpected power outage.
- Reconnect power to restart the board.
- **Alternatively:**

    Press the Restart button on the STM32 board to simulate a power cycle without physically disconnecting power.

Expected Behavior on Reboot

- Upon reboot, the program should retrieve the saved `read_addr`, `write_addr`, and `buffer_len` states from FRAM.
- Pressing the **Restart** button should add additional test data to the FRAM buffer, confirming that data integrity and buffer states are preserved across power cycles.

3. Unit Test

- The `test_LoadSaveBufferState` function verifies that `read_addr` and `write_addr` in FRAM accurately reflect buffer states before and after simulated power outages.

- This unit test ensures that the buffer state is correctly saved and reloaded, maintaining data continuity across restarts.

**Additional Context**
The unit test for this feature validates the implemented methods at a software level when the "Restart" button is clicked. It ensures that the buffer state (read_addr and write_addr) accurately reflects data continuity across simulated restarts.

For a full testing procedure, including power cycle simulation, please refer to the Test Procedure section.

**Task List**


**Relevant Issues**
Closes #69
